### ISSUE TITLE
Define DIDComm for gRPC

### ISSUE THREAD

**Issue by tmarkovski** [decentralized-identity/didcomm-messaging#113]
## Motivation

Currently, DIDComm defines a JSON message structure and a communication method for exchange over REST. Should we also define a way to use DIDComm over gRPC? As it currently stands, DIDComm can be used over gRPC due to compatiblity with JSON, but this doesn't utilize the gRPC binary serialization nor the benefits of HTTP/2.
gRPC offers some pretty cool features inherent to HTTP/2 like bi-directional message streaming over a single TCP connection. This can provide a very powerful pattern where didcomm can benefit to expand on the single request/response interaction. For example, the following exchange can be done over a single TCP connection:

```
client -> server: credential request
server -> client: request presentation
client -> server: submit presentation
server -> client: issue credential
server -> client: basic message
```

Even better, the client and server can establish an ephemeral pairwise connection for that exchange only, for example using did:key method key exchange. This kind of behavior adds flexibility to connections where they can used for temporary secure exchange, in addition to being more permanent, which is how they are today.

## How to structure DIDComm over gRPC

Message contracts in DIDComm can be expressed using proto syntax. For example:

Routing forward message in JSON
```
{
    "@type": "https://didcomm.org/routing/2.0/forward",
    "to"   : "did:foo:1234abcd",
    "payloads~attach": [
        // One payload
    }
}
```

Routing forward message using proto syntax
```
message Forward {
    string to = 1;
    repeated Attachment Payload = 2 [json_name="payloads~attach"];
}
message Attachment {
    string id = 1;
    string contentType = 2;
    byte payload = 3;
}
```

Service endpoints can also be expressed using
```
service DidcommService {
    rpc Endpoint(Any) return (Any);

    rpc StreamEndpoint(Stream) returns (Stream);

    rpc JsonEndpoint(Struct) returns (Struct);
}
```

The above service defines three different endpoints for added flexibility and compatibility. Clients that can only communicate using JSON, can use `JsonEndpoint`, clients that do not have ability to use HTTP2 streaming, can use `Endpoint`, etc.

This repo could potentially maintain `.proto` files defined for each RFC and allow extensions to be developed. For example, credential exchange, presentations, etc.

One of the benefits of using protocol buffers is also the broad availability of languages that can use these contracts. Protobuf generates native code and message structures for both messages and services. This will help implementers get started immediately, and not worry about meeting correct structure for interop. Some applications may want to use a simple feature of DIDComm, like forward message. They wouldn't need to adopt a full code framework to that. They can just import the proto files and prepare a call.

I'm curious how the community feels about this. I'd be happy to discuss this in a WG call.

**Comment by downTheFallLine** [comment]
There's a lot of goodness in gRPC and has found traction in a number of projects.  No doubt about it.   

I think there's a couple of ways to look at this.   

One could explicitly include gRPC language into the didcomm spec itself.   And do that for other messaging layers (what about X.12?, etc. etc.).  In my mind, that leads to a very "heavy weight" specification, that over time becomes more difficult to amend/modify, and more difficult for people to adopt.  An example of this strategy was the development of Asynchronous Transfer Mode Switched Virtual Circuits.   

On the other hand, a set of informative companion documents could be created that put didcomm into a particular context: e.g., didcomm bindings for gRPC (or x.12).  This approach allows the precious bandwidth of the design team to focus on the essential aspects of the protocol, results in a "light weight" protocol that does NOT preclude the use of various transports such as gRPC.

Some food for thought...

**Comment by tmarkovski** [comment]
@downTheFallLine I like option 2. What things about describing use of other protocols should this include? If DIDComm is a transport agnostic protocol, is there anything about the protocol extensions that the core RFCs should be concerned about? Things like, type URIs must be preserved? DID Service definition should be published, etc? Other than gRPC, I'm also considering Zeroconf/Bonjour as a transport layer for peer to peer offline (internet-less) communication.

**Comment by downTheFallLine** [comment]
@tmarkovski i think you're describing the binding.  Indeed, if you keep the spec itself light weight, there needs to be other documents that describe particular implementations (gRPC running over barbed-wire transport (kidding)).    Seriously, if memory serves, this was the approach taken in the early days of TCP, before ethernet was THE standard (i remember getting involved in the ill-fated IP/ATM effort :) 

So yes, keep the spec itself light.   Develop other documents as a best practice for particular implementations, in my view.

**Comment by TelegramSam** [comment]
I think there are to possible directions here: include gRPC as a transport in the spec, or produce an independent document that contains the same material. The location of the 'profile' or application standard depends quite a bit on how widespread it is, and if this is something that can be used directly from a DID Document, or is a transport that can be upgraded from a normal DIDComm connection.

I'll add this to the agenda, though I'm not very confident that we'll get to it this week. It might be next week.

**Comment by dhh1128** [comment]
I think that if we supported gRPC as a canonical representation of DIDComm messages, we'd need to define DIDComm with an abstract data model. See [my comment in #112 about the pros and cons of this](https://github.com/decentralized-identity/didcomm-messaging/issues/112#issuecomment-696209805).

Could we take an alternate approach and make gRPC an extension or transformation, rather than baking abstraction into the core of DIDComm?

**Comment by llorllale** [comment]
> Could we take an alternate approach and make gRPC an extension or transformation

I like this approach.

**Comment by TelegramSam** [comment]
I like this as an extension as well.

**Comment by tmarkovski** [comment]
Sure. What does an extension look like? Does it define it's own data model and protocol? What makes an extension of DIDComm an extension? Which core properties of DIDComm must it support to be considered an addition or extension to the core protocol?

**Comment by TelegramSam** [comment]
This is all the spec currently says about extension: https://identity.foundation/didcomm-messaging/spec/#extensions

If we need clarifications added, we should discuss the language. This is on the agenda for today's call.

**Comment by TelegramSam** [comment]
Related to #222

**Comment by TelegramSam** [comment]
PR in to add requirements to the spec per #222

**Comment by TelegramSam** [comment]
Needs some updates to the Extensions text to clarify how this is to be done.

**Comment by TelegramSam** [comment]
Definition on how to add a transport as an extension was included in #320. Closing this, as we've seen no contributions on this topic. Please reopen if interest grows in gRPC as a transport.
